retrieval/discovery/azure.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// createAzureClient is a helper function for creating an Azure compute client to ARM.
+func createAzureClient(cfg config.AzureSDConfig) (azureClient, error) {
+	var c azureClient
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(cfg.TenantID)
+	if err != nil {
+		return azureClient{}, err
+	}
+	spt, err := azure.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, cfg.ClientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return azureClient{}, err
+	}
+
+	c.vm = compute.NewVirtualMachinesClient(cfg.SubscriptionID)
+	c.vm.Authorizer = spt
+
+	c.nic = network.NewInterfacesClient(cfg.SubscriptionID)
+	c.nic.Authorizer = spt
+
+	return c, nil
+}
+
+func (ad *AzureDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{}
+	client, err := createAzureClient(*ad.cfg)
+	if err != nil {
+		return tg, fmt.Errorf("could not create Azure client: %s", err)
+	}
+
+	var machines []compute.VirtualMachine
+	result, err := client.vm.ListAll()
+	if err != nil {
+		return tg, fmt.Errorf("could not list virtual machines: %s", err)
+	}
+	machines = append(machines, *result.Value...)
+
+	// If we still have results, keep going until we have no more.
+	for result.NextLink != nil {
+		result, err = client.vm.ListAllNextResults(result)
+		if err != nil {
+			return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		}
+		machines = append(machines, *result.Value...)
+	}
+	log.Debugf("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// We have the slice of machines. Now turn them into targets.
+	// Doing them in go routines because the network interface calls are slow.
+	var wg sync.WaitGroup
+	targets := make([]model.LabelSet, len(machines))
+	finished := make(chan bool, 1)
+	errors := make(chan error, 0)
+	for i, vm := range machines {
+		wg.Add(1)
+		go func(i int, vm compute.VirtualMachine) {
+			defer wg.Done()
+			resourceGroup := strings.ToLower(strings.Split(*vm.ID, "/")[4])
+			labels := model.LabelSet{
+				azureLabelMachineID:            model.LabelValue(*vm.ID),
+				azureLabelMachineName:          model.LabelValue(*vm.Name),
+				azureLabelMachineLocation:      model.LabelValue(*vm.Location),
+				azureLabelMachineResourceGroup: model.LabelValue(resourceGroup),
+			}
+
+			if vm.Tags != nil {
+				for k, v := range *vm.Tags {
+					name := strutil.SanitizeLabelName(k)
+					labels[azureLabelMachineTag+model.LabelName(name)] = model.LabelValue(*v)
+				}
+			}
+
+			// Get the IP address information via seperate call to the network provider.
+			for _, nic := range *vm.Properties.NetworkProfile.NetworkInterfaces {
+				group := strings.Split(*nic.ID, "/")[4]
+				name := strings.Split(*nic.ID, "/")[8]
+				networkInterface, err := client.nic.Get(group, name, "")
+
+				// If we can't get the network interface, send an Error to the errors channel.
+				if err != nil {
+					log.Errorf("Unable to get network interface %s: %s", name, err)
+					errors <- err
```

Comment:
If there's more than one error, the other goroutines will block here indefinitely as the errors channel is unbuffered.

What'd be better is to have the goroutine return everything on the channel, so either an error or the target, and collect them up. This will also remove the concurrent writes to the `targets` map which is not safe.